application.go

```diff
@@ -57,3 +64,120 @@ func (ac *ApplicationClient) GetApplication(
 
 	return res.Application, nil
 }
+
+// GetApplicationsDelegatingToGateway returns the application addresses that are
+// delegating to the given gateway address.
+func (ac *ApplicationClient) GetApplicationsDelegatingToGateway(
+	ctx context.Context,
+	gatewayAddress string,
+	queryHeight uint64,
+) ([]string, error) {
+	allApplications, err := ac.GetAllApplications(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("GetApplicationsDelegatingToGateway: error getting all applications: %w", err)
+	}
+
+	gatewayDelegatingApplications := make([]string, 0)
+	for _, application := range allApplications {
+		appRing := ApplicationRing{Application: application}
+		// Get the gateways that are delegated to the application
+		// at the query height and check if the given gateway address is in the list.
+		gatewaysDelegatedTo := appRing.ringAddressesAtBlock(queryHeight)
+		if slices.Contains(gatewaysDelegatedTo, gatewayAddress) {
+			// The application is delegating to the given gateway address, add it to the list.
+			gatewayDelegatingApplications = append(gatewayDelegatingApplications, application.Address)
+		}
+	}
+
+	return gatewayDelegatingApplications, nil
+}
+
+type ApplicationRing struct {
```

Comment:
Not a strong opinion but I believe `ApplicationRing` should embed a block client just like it has a `PublicKeyFetcher`, so `GetRing` could use it to get the query height instead of having it passed to.

I see no use case where we want to sign the relay request for a block that is different from the current one.

This will also remove the need of `signer.Sign` to pass in the query height since it already has `ApllicationRing` as an arg.

Reply:
Thank you for the review.
`Not a strong opinion but I believe ApplicationRing should embed a block client just like it has a PublicKeyFetcher, so GetRing could use it to get the query height instead of having it passed to.`

This is done to minimize the potential calls made to nodes directly by the function. For a more complex use case, the `PublicKeyFetcher` can be a struct that has prefetched all public key, address pairs from a POKT node.

`I see no use case where we want to sign the relay request for a block that is different from the current one.`

Thank you for pointing this out: I have updated the ApplicationRing struct to include a `CurrentHeight` field.

`This will also remove the need of signer.Sign to pass in the query height since it already has ApllicationRing as an arg.`

Thank you for the review, the above update fixed this as well.